added `cluster` function to `list` module

massivefermion opened the pull request:
I think this could be useful, as I just ran into a use case.
But I'm not sure about the implementation so hopefully we can make it better.

massivefermion commented:
I just realized you could do this by doing a `chunk` and then a `filter` but maybe that means traversing the list twice when you could just do it once!

lpil commented:
This seems like an unusual function to have in the stdlib to me. Which other stdlibs have this one?

massivefermion commented:
I couldn't find any, so you're probably right. I just thought it may be useful.
I was trying some different methods for to parse some strings and one way was to use `regex.split` and it returns a list of strings. If you capture multiple matches based on a single regex using `regex.split`, that list of strings will have the captured submatches  next to each other in the array. A function like `cluster` could actually turn that list into a list of lists and each of those lists contains each of the submatches for each match. 
But I understand if this use case seems to niche and doesn't justify adding a function to stdlib.

lpil commented:
Cool, ok. I think it's likely a bit too niche, if someone wants this they could use `split_while` which is quite similar.
